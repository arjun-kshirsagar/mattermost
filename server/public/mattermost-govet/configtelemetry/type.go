// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package configtelemetry

import (
	"go/ast"
	"go/token"
	"strings"
)

var ignoreComments = []string{
	"telemetry: none",
	"Deprecated: do not use",
}

// typeFieldMap will generate map of its all possible selectors and adds them as a key. The value
// is the position of the declaration in the type spec.
func typeFieldMap(file *ast.File, typeName string) (map[string]token.Pos, error) {

	var fieldsMap map[string]token.Pos

	ast.Inspect(file, func(n ast.Node) bool {
		decl, ok := n.(*ast.GenDecl)
		if !ok || decl.Tok != token.TYPE || len(decl.Specs) < 1 {
			return true
		}

		spec, ok := decl.Specs[0].(*ast.TypeSpec)
		if !ok || spec.Name.Name != typeName {
			return true
		}

		typ, ok := spec.Type.(*ast.StructType)
		if !ok {
			return true
		}

		fieldsMap = fields(typ)

		return false
	})

	return fieldsMap, nil
}

func fields(str *ast.StructType) map[string]token.Pos {
	structFields := make(map[string]token.Pos)

	for _, field := range str.Fields.List {
		if field.Comment != nil && ignored(field.Comment) {
			continue
		}

		expr := field.Type
		if ptr, ok := field.Type.(*ast.StarExpr); ok {
			expr = ptr.X
		}

		if len(field.Names) != 1 {
			panic("unhandled struct field definition")
		}

		key := field.Names[0].Name

		v, ok := expr.(*ast.Ident)
		if !ok || v.Obj == nil || v.Obj.Decl == nil {
			structFields[key] = field.Pos()
			continue
		}

		typeDecl, ok := v.Obj.Decl.(*ast.TypeSpec)
		if !ok {
			continue
		}

		typ, ok := typeDecl.Type.(*ast.StructType)
		if !ok {
			continue
		}

		for k, v := range fields(typ) {
			structFields[key+"."+k] = v
		}
	}

	return structFields
}

func ignored(comments *ast.CommentGroup) bool {
	for _, comment := range comments.List {
		for _, ignore := range ignoreComments {
			if strings.Contains(comment.Text, ignore) {
				return true
			}
		}
	}
	return false
}
