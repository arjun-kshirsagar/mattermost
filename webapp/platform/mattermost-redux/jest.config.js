// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

/** @type {import('jest').Config} */

module.exports = {
    clearMocks: true,
    moduleNameMapper: {
        '^@mattermost/types/(.*)$': '<rootDir>/../types/src/$1',
        '^mattermost-redux/(.*)$': '<rootDir>/src/$1',
        '^.+\\.(jpg|jpeg|png|apng|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$':
            'identity-obj-proxy',
        '^.+\\.(css|less|scss)$': 'identity-obj-proxy',
        '^.*i18n.*\\.(json)$': '<rootDir>/../../channels/src/tests/i18n_mock.json',
    },
    // setupFilesAfterEnv: ['<rootDir>/test/setup_jest.ts'],
    // testEnvironment: 'jsdom',
};
