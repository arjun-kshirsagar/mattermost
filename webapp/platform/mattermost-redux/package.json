{
  "name": "mattermost-redux",
  "version": "9.3.0",
  "description": "Redux utilities for working with Mattermost",
  "keywords": [
    "mattermost"
  ],
  "homepage": "https://github.com/mattermost/mattermost/tree/master/webapp/platform/mattermost-redux#readme",
  "license": "MIT",
  "files": [
    "lib"
  ],
  "exports": {
    "./*": "./lib/*"
  },
  "typesVersions": {
    ">=3.1": {
      "*": [
        "./lib/*",
        "./lib/index.*"
      ]
    }
  },
  "repository": {
    "type": "git",
    "url": "github:mattermost/mattermost",
    "directory": "webapp/platform/mattermost-redux"
  },
  "dependencies": {
    "lodash": "^4.17.21",
    "moment-timezone": "^0.5.38",
    "redux": "^4.2.0",
    "redux-batched-actions": "^0.5.0",
    "redux-thunk": "^2.4.2",
    "rudder-sdk-js": "^1.0.16",
    "shallow-equals": "^1.0.0",
    "timezones.json": "^1.6.1"
  },
  "devDependencies": {
    "@types/jest": "28.1.8",
    "identity-obj-proxy": "3.0.0",
    "isomorphic-fetch": "3.0.0",
    "jest": "27.1.0",
    "jest-environment-jsdom": "29.7.0",
    "nock": "13.2.8",
    "@redux-devtools/extension": "3.2.3",
    "redux-mock-store": "1.5.4",
    "typescript": "^5.0.0"
  },
  "peerDependencies": {
    "@mattermost/types": "9.3.0",
    "@mattermost/client": "9.3.0",
    "typescript": "^4.3.0 || ^5.0.0"
  },
  "peerDependenciesMeta": {
    "typescript": {
      "optional": true
    }
  },
  "scripts": {
    "build": "tsc --build tsconfig.build.json --verbose",
    "check": "eslint --ext .js,.jsx,.tsx,.ts ./src --quiet",
    "run": "tsc --watch --preserveWatchOutput",
    "test": "cross-env TZ=Etc/UTC jest",
    "test-ci": "cross-env TZ=Etc/UTC jest --ci --forceExit --detectOpenHandles --maxWorkers=100%",
    "clean": "rm -rf lib node_modules *.tsbuildinfo"
  }
}
