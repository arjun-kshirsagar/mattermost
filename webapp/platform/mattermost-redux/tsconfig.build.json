{
    "compilerOptions": {
        "module": "commonjs",
        "moduleResolution": "node",
        "target": "es2022",
        "declaration": true,
        "strict": true,
        "resolveJsonModule": true,
        "isolatedModules": true,
        "esModuleInterop": true,
        "allowSyntheticDefaultImports": true,
        "jsx": "react",
        "outDir": "./lib",
        "rootDir": "./src",
        "composite": true,
        "useUnknownInCatchVariables": false,
        "paths": {
            "mattermost-redux/*": ["./src/*"]
        }
    },
    "include": [
        "./src/**/*"
    ],
    "exclude": [
        "**/*.test.*"
    ],
    "references": [
        {"path": "../client"},
        {"path": "../types"}
    ]
}
