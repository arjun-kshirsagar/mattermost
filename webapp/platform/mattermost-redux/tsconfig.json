{
    "compilerOptions": {
        "allowJs": true,
        "module": "commonjs",
        "moduleResolution": "node",
        "target": "es2022",
        "declaration": true,
        "strict": true,
        "resolveJsonModule": true,
        "noEmit": true,
        "isolatedModules": true,
        "esModuleInterop": true,
        "allowSyntheticDefaultImports": true,
        "jsx": "react",
        "outDir": "./lib",
        "rootDir": "./src",
        "composite": true,
        "useUnknownInCatchVariables": false,
        "paths": {
            "mattermost-redux/*": ["./src/*"]
        }
    },
    "include": [
        "./src/**/*"
    ],
    "references": [
        {"path": "../client"},
        {"path": "../types"}
    ]
}
