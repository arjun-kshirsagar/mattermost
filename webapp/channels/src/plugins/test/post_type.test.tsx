// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {ComponentProps} from 'react';
import React from 'react';
import {screen} from '@testing-library/react';

import {Preferences} from 'mattermost-redux/constants';

import PostMessageView from 'components/post_view/post_message_view/post_message_view';
import {renderWithContext} from 'tests/react_testing_utils';

// Mock the ShowMore component to avoid context issues
jest.mock('components/post_view/show_more', () => () => <div data-testid="show-more">Show More</div>);

const PostTypePlugin = () => (
    <span data-testid='pluginId'>{'PostTypePlugin'}</span>
);

describe('plugins/PostMessageView', () => {
    const post = {type: 'testtype', message: 'this is some text', id: 'post_id'} as any;

    const requiredProps: ComponentProps<typeof PostMessageView> = {
        post,
        pluginPostTypes: {
            testtype: {
                id: 'some id',
                pluginId: 'some plugin id',
                component: PostTypePlugin,
                type: '',
            },
        },
        theme: Preferences.THEMES.denim,
        enableFormatting: true,
        currentRelativeTeamUrl: 'team_url',
    };

    test('should render properly with extended post type', () => {
        const {container} = renderWithContext(
            <PostMessageView {...requiredProps}/>,
        );

        // Testing that the custom plugin component renders
        expect(screen.getByTestId('pluginId')).toBeInTheDocument();
        expect(screen.getByText('PostTypePlugin')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    test('should render properly with no extended post type', () => {
        const props = {...requiredProps, pluginPostTypes: {}};
        const {container} = renderWithContext(
            <PostMessageView {...props}/>,
        );

        // When no plugin post type is available, it should not render the plugin component
        expect(screen.queryByTestId('pluginId')).not.toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
