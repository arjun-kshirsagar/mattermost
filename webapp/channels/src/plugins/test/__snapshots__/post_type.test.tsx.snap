// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/PostMessageView should render properly with extended post type 1`] = `
<div>
  <span
    data-testid="pluginId"
  >
    PostTypePlugin
  </span>
</div>
`;

exports[`plugins/PostMessageView should render properly with no extended post type 1`] = `
<div>
  <div
    data-testid="show-more"
  >
    Show More
  </div>
</div>
`;
